Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -133,6 +143,103 @@ void WebExtension::populateDisplayStringsIfNeeded()
         recordError(createError(Error::InvalidDescription));
 }
 
+const String& WebExtension::contentSecurityPolicy()
+{
+    populateContentSecurityPolicyStringsIfNeeded();
+    return m_contentSecurityPolicy;
+}
+
+void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+{
+    if (m_parsedManifestContentSecurityPolicyStrings)
+        return;
+
+    m_parsedManifestContentSecurityPolicyStrings = true;
+
+    RefPtr manifestObject = this->manifestObject();
+    if (!manifestObject)
+        return;
+
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy
+
+    if (supportsManifestVersion(3)) {
+        if (RefPtr policyObject = manifestObject->getObject(contentSecurityPolicyManifestKey)) {
+            m_contentSecurityPolicy = policyObject->getString(contentSecurityPolicyExtensionPagesManifestKey);
+            if (!m_contentSecurityPolicy && (!policyObject->size() || policyObject->getValue(contentSecurityPolicyExtensionPagesManifestKey)))
+                recordError(createError(Error::InvalidContentSecurityPolicy));
+        }
+    } else {
+        m_contentSecurityPolicy = manifestObject->getString(contentSecurityPolicyManifestKey);
+        if (!m_contentSecurityPolicy && manifestObject->getValue(contentSecurityPolicyManifestKey))
+            recordError(createError(Error::InvalidContentSecurityPolicy));
+    }
+
+    if (!m_contentSecurityPolicy)
+        m_contentSecurityPolicy = "script-src 'self'"_s;
+}
+
+bool WebExtension::hasOptionsPage()
+{
+    populatePagePropertiesIfNeeded();
+    return !m_optionsPagePath.isEmpty();
+}
+
+bool WebExtension::hasOverrideNewTabPage()
+{
+    populatePagePropertiesIfNeeded();
+    return !m_overrideNewTabPagePath.isEmpty();
+}
+
+const String& WebExtension::optionsPagePath()
+{
+    populatePagePropertiesIfNeeded();
+    return m_optionsPagePath;
+}
+
+const String& WebExtension::overrideNewTabPagePath()
+{
+    populatePagePropertiesIfNeeded();
+    return m_overrideNewTabPagePath;
+}
+
+void WebExtension::populatePagePropertiesIfNeeded()
+{
+    if (m_parsedManifestPageProperties)
+        return;
+
+    m_parsedManifestPageProperties = true;
+
+    RefPtr manifestObject = this->manifestObject();
+    if (!manifestObject)
+        return;
+
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/options_ui
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/options_page
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/chrome_url_overrides
+
+    RefPtr optionsDictionary = manifestObject->getObject(optionsUIManifestKey);
+    if (optionsDictionary) {
+        m_optionsPagePath = optionsDictionary->getString(optionsUIPageManifestKey);
```

Comment:
Sorry didn't catch this earlier but …Object would be a better suffix for these now.
```suggestion
    if (RefPtr optionsObject = manifestObject->getObject(optionsUIManifestKey)) {
        m_optionsPagePath = optionsObject->getString(optionsUIPageManifestKey);
```